firefox_puppeteer/ui/tabbar.py

```diff
@@ -3,126 +3,315 @@
-                                         ' found'.format(target))
+            self.marionette.switch_to_window(start_handle)
+            raise errors.UnknownTabError("No tab found for '{}'".format(target))
+
+        raise ValueError("The 'target' parameter must either be an index or a callable")
+
+    @staticmethod
+    def get_handle_for_tab(marionette, tab_element):
+        """Retrieves the marionette handle for the given :clas:`Tab` instance.
+
+        :param marionette: An instance of the Marionette client.
 
-        raise TypeError("Invalid type for 'target': {}".format(type(target)))
+        :param tab_element: The DOM element corresponding to a tab inside the tabs toolbar.
 
-    def switch_to_tab(self, tab):
+        :returns: `handle` of the tab
         """
-        Switch to (activate) the specified tab.
+        # TODO: This introduces coupling with marionette's window handles
+        # implementation. To avoid this, the capacity to get the XUL
+        # element corresponding to the active window according to
+        # marionette or a similar ability should be added to marionette.
+        handle = marionette.execute_script("""
+          let win = arguments[0].linkedBrowser.contentWindowAsCPOW;
+          return win.QueryInterface(Ci.nsIInterfaceRequestor)
+                    .getInterface(Ci.nsIDOMWindowUtils)
+                    .outerWindowID.toString();
+        """, script_args=[tab_element._tab])
+
+        return handle
+
+
+class Tab(UIBaseLib):
+    """Wraps a tab DOM element."""
+
+    def __init__(self, marionette_getter, window, tab_element):
+        UIBaseLib.__init__(self, marionette_getter, window)
+
+        self._tab = tab_element
+        self._handle = TabBar.get_handle_for_tab(self.marionette, self)
 
-        :param tab: Either a :class:`TabElement`, an index of `tabs` or a
-                    substring of the label.
+        # Ensure the tab has been fully loaded
+        Wait(self.marionette).until(
+            lambda mn: mn.execute_script("""
+              return !arguments[0].hasAttribute('busy');
+            """, script_args=[tab_element])
+        )
+
+    def __eq__(self, other):
+        return self.handle == other.handle
+
+    @property
+    def handle(self):
+        """The `handle` of the content window.
+
+        :returns: content window `handle`
         """
-        if not isinstance(tab, HTMLElement):
-            tab = self.get_tab(tab)
-        return tab.click()
+        return self._handle
+
+    @property
+    def close_button(self):
+        """The DOM element which represents the tab close button.
 
-    class TabElement(DOMElement):
+        :returns: Reference to the tab close button
         """
-        Wraps a tab element.
+        return self.tab.find_element('anon attribute', {'anonid': 'close-button'})
+
+    @property
+    def tab(self):
+        """The inner tab DOM element.
+
+        :returns: Tab DOM element
         """
+        return self._tab
```

Comment:
I know and I wasn't that sure what to use here either. :) But sounds like a good compromise, which we should also do for the Window classes.